src/Compilers/VisualBasic/Portable/VisualBasicParseOptions.vb

```diff
@@ -174,16 +187,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         ''' <param name="version">The parser language version.</param>
         ''' <returns>A new instance of VisualBasicParseOptions if different language version is different; otherwise current instance.</returns>
         Public Shadows Function WithLanguageVersion(version As LanguageVersion) As VisualBasicParseOptions
-            version = version.MapLatestToVersion()
-            If version = _languageVersion Then
+            If version = _specifiedLanguageVersion Then
                 Return Me
             End If
 
-            If Not version.IsValid Then
+            Dim effectiveVersion = version.MapSpecifiedToEffectiveVersion()
+            If Not effectiveVersion.IsValid Then
                 Throw New ArgumentOutOfRangeException(NameOf(version))
             End If
 
-            Return New VisualBasicParseOptions(Me) With {._languageVersion = version}
+            Return New VisualBasicParseOptions(Me) With {._specifiedLanguageVersion = version, ._languageVersion = effectiveVersion}
```

Comment:
Eg copy the constructor `VisualBasicParseOptions.New(other As VisualBasicParseOptions)` but with two additional parameters.
`,SpecifiedVersion As LanguageVersion, EffectiveVersion As LanguageVersion)` 

This will save use from recomputing the already computed version mapping.